src/_document/render.tsx

```diff
@@ -22,17 +27,36 @@ export default async function renderDocument({
     ? customDocumentFile.default
     : ((NextDocument as unknown) as DocumentType);
 
-  // TODO: this should get updated by HeadManagerContext
-  // See: https://github.com/vercel/next.js/blob/2790ea411b8b0bf2be627a5cc5ba77626079c46d/packages/next/next-server/server/render.tsx#L563
-  // Problem, can we get this to evaluate to false: https://github.com/vercel/next.js/blob/2790ea411b8b0bf2be627a5cc5ba77626079c46d/packages/next/next-server/lib/side-effect.tsx#L3?
-  const head: JSX.Element[] = [];
+  let head: JSX.Element[] = [];
 
   const renderPage: RenderPage = () => {
-    // Render markup that will be injected into #__next element by Next.JS in the Main component of Document
-    const html = renderToStaticMarkup(pageElement);
+    const html = renderToStaticMarkup(
+      <HeadManagerContext.Provider
+        value={{
+          updateHead: (state) => {
+            head = state;
+          },
+          mountedInstances: new Set(),
+        }}
+      >
+        {pageElement}
+      </HeadManagerContext.Provider>
+    );
     return { html, head };
   };
 
+  // https://github.com/vercel/next.js/blob/canary/packages/next/pages/_document.tsx#L524
+  // Default behaviour of Main component is to dangerouslySetInnerHTML with the html
+  // string rendered above. This works, but will break all client side interactions
+  // as event handlers are lost in static markup
+  // @ts-ignore
+  documentModule.Main = () => {
+    const { inAmpMode, docComponentsRendered } = useContext(DocumentContext);
+    docComponentsRendered.Main = true;
+    if (inAmpMode) return <>{AMP_RENDER_TARGET}</>;
```

Comment:
Are these 3 lines (54,55,56) necessary for the scope of our tests?
I can remove them and still have green tests.

Reply:
They are not needed, but wanted to have an implementation that matches NextJS

Reply:
I see. In this case I'd friendly suggest to:
- make clear in a comment that those lines do nothing
- add an `/* istanbul ignore next */` before line 56 to restore code coverage to 100%

I'm insisting in keeping test coverage to 100% to pre-catch some of the many regression bug this project will suffer.